rts/motoko-rts/src/memory/ic.rs

```diff
@@ -1,66 +1,90 @@
 /// `Memory` implementation allocates in Wasm heap with Wasm `memory.grow` instruction.
 pub struct IcMemory;
 
-impl Memory for IcMemory {
-    #[inline]
-    unsafe fn alloc_words(&mut self, n: Words<usize>) -> Value {
-        crate::gc::incremental::get_partitioned_heap().allocate(self, n)
-    }
-
-    /// Page allocation. Ensures that the memory up to, but excluding, the given pointer is allocated.
-    /// Ensure a memory reserve for the incremental GC.
-    #[inline(never)]
-    unsafe fn grow_memory(&mut self, ptr: usize) {
-        const LAST_PAGE_LIMIT: usize = 0xFFFF_FFFF_FFFF_0000;
-        debug_assert_eq!(LAST_PAGE_LIMIT, usize::MAX - WASM_PAGE_SIZE.as_usize() + 1);
-        debug_assert!(memory_reserve() <= MAXIMUM_MEMORY_SIZE.as_usize());
-        let limit = min(
-            LAST_PAGE_LIMIT,
-            MAXIMUM_MEMORY_SIZE.as_usize() - memory_reserve(),
-        );
-        if ptr > limit {
-            rts_trap_with("Cannot grow memory")
-        };
-        let page_size = WASM_PAGE_SIZE.as_usize();
-        let total_pages_needed = ((ptr + page_size - 1) / page_size) as usize;
-        let current_pages = wasm64::memory_size(0);
-        if total_pages_needed > current_pages {
-            if wasm64::memory_grow(0, total_pages_needed - current_pages) == core::usize::MAX {
-                // replica signals that there is not enough memory
-                rts_trap_with("Cannot grow memory");
-            }
-        }
-    }
+#[enhanced_orthogonal_persistence]
+pub fn wasm_memory_size() -> usize {
+    core::arch::wasm64::memory_size(0)
 }
 
-#[no_mangle]
-unsafe extern "C" fn get_reclaimed() -> Bytes<usize> {
-    crate::gc::incremental::get_partitioned_heap().reclaimed_size()
+#[classical_persistence]
+pub fn wasm_memory_size() -> usize {
+    core::arch::wasm32::memory_size(0)
 }
 
-#[no_mangle]
-pub unsafe extern "C" fn get_total_allocations() -> Bytes<usize> {
-    get_heap_size() + get_reclaimed()
+#[enhanced_orthogonal_persistence]
+pub fn wasm_memory_grow(pages: usize) -> usize {
+    core::arch::wasm64::memory_grow(0, pages)
 }
 
-#[no_mangle]
-pub unsafe extern "C" fn get_heap_size() -> Bytes<usize> {
-    crate::gc::incremental::get_partitioned_heap().occupied_size()
+#[classical_persistence]
+pub fn wasm_memory_grow(pages: usize) -> usize {
+    core::arch::wasm32::memory_grow(0, pages)
 }
 
-#[no_mangle]
-pub unsafe extern "C" fn get_max_live_size() -> Bytes<usize> {
-    crate::gc::incremental::get_max_live_size()
+/// Page allocation. Ensures that the memory up to, but excluding, the given pointer is allocated.
+/// Ensure a memory reserve of at least one Wasm page depending on the canister state.
+/// `memory_reserve`: A memory reserve in bytes ensured during update and initialization calls.
+//  For use by queries and upgrade calls. The reserve may vary depending on the GC and the phase of the GC.
+unsafe fn grow_memory(ptr: u64, memory_reserve: usize) {
```

Comment:
Yes, this is used in both 32-bit and 64-bit. In 32-bit it is an optimization (w.r.t. to the IC cost model) to avoid  overflows in `alloc_words` for linear memory (classical GCs only). I thought to keep that logic, although it looks asymmetric.

```
let old_hp = getHP() as u64;
let new_hp = old_hp + delta;
```